proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+.. _`#291`: https://github.com/ghc-proposals/ghc-proposals/pull/291
+.. _`#378`: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst
+.. _`#402`: https://github.com/ghc-proposals/ghc-proposals/pull/402
+.. _`#420`: https://github.com/ghc-proposals/ghc-proposals/pull/420
+.. _Type Variables in Patterns: https://richarde.dev/papers/2018/pat-tyvars/pat-tyvars.pdf
+.. _Kind Inference for Datatypes: https://richarde.dev/papers/2020/kind-inference/kind-inference.pdf
+.. _`Haskell 2010 Report`: https://www.haskell.org/onlinereport/haskell2010/haskellch10.html
+
+* `#126`_: Accepted, implemented proposal on accepting type arguments to constructor
+  patterns, allowing constructions like ``f (Just @Int x) = x + 5``
+  and ``g (Dynamic @a x) = show (typeOf (x :: a))``.
+* `#128`_: Accepted, implemented proposal that allows type variables in certain contexts
+  to refer to *types*, not just other type variables. Example: ``f (True :: b) = (False :: b)``;
+  here, ``b`` becomes a synonym for ``Bool``. (Previously, a type variable was required to remain
+  abstract.)
+* `#155`_: Accepted, not implemented proposal on accepting type arguments to
+  lambdas, allowing constructions like ``\ @a (x :: a) -> x``.
+* `#238`_: Not yet accepted proposal that updates and extends `#155`_ to
+  include a reshuffling of extensions around scoped type variables, as well
+  as describing clearer rules mediating between the old ``-XScopedTypeVariables``
+  bahavior and the new proposed behavior.
+* `#285`_: Accepted, not implemented proposal introducing ``-XNoImplicitForAll``
+  and ``-XNoPatternSignatureBinds``, restricting where type variables may implicitly
+  be brought into scope.
+* `#291`_: Not yet accepted proposal that is an amendment to `#126`_ to change
+  the way type variables are brought into scope to be more like term variables,
+  and less like pattern signatures.
+* `#420`_: Not yet accepted proposal that is an amendment to `#285`_, with a
+  very tiny, technical tweak to a definition.
+
+This proposal supersedes all of the proposals above, including the accepted
+ones. The goal of writing this is to provide a unified framework, so that
+we can design our language with cohesive treatment of scoped type variables,
+instead of simply an agglomeration of features.
+
+Motivation
+----------
+
+With GHC's powerful type-level programming features, we need powerful abilities
+to bring type variables into scope. The proposal defers to the individual proposals
+linked above for motivation for why we generally want these type-level features.
+Individual aspects of this unifying proposal are motivated near where they are
+introduced.
+
+Principles
+----------
+
+The design here is guided by several high-level principles. These principles are not
+sacro-sanct, such that they admit no exceptions. Instead, we strive to uphold these
+principles, but relax them when doing so is well motivated.
```

Comment:
Maybe, given the current state of play, you shouldn't call these `X Principle` when you don't intend to have them figure in the _Design principles_ document?

Reply:
Ah, but I do! This proposal predates the Principles document, #452. Once that's accepted, I'll restructure this PR to move the principles there.